File: nhitomi-idx/nhitomi/Discord/DestroyTrigger.cs
using System.Threading;
using System.Threading.Tasks;
using Discord;

namespace nhitomi.Discord
{
    /// <summary>
    /// A trigger that deletes a message and destroys its interactive.
    /// </summary>
    public class DestroyTrigger : ReactionTrigger
    {
        readonly InteractiveMessage _message;

        public override IEmote Emote => new Emoji("\uD83D\uDDD1"); // trashcan

        public DestroyTrigger(InteractiveMessage message)
        {
            _message = message;
        }

        protected override async Task<bool> RunAsync(CancellationToken cancellationToken = default)
        {
            try
            {
                await _message.Reply.DeleteAsync();
                await _message.DisposeAsync();

                return true;
            }
            catch
            {
                return false;
            }
        }
    }
}

File: nhitomi-discord/src/Triggers/DestroyTrigger.ts
import { ReactionTrigger } from "../Interactive/trigger";
import { InteractiveMessage } from "../Interactive/message";

export class DestroyTrigger extends ReactionTrigger {
  constructor(interactive: InteractiveMessage) {
    super(interactive, "\uD83D\uDDD1");
  }

  protected async run(): Promise<boolean> {
    // setTimeout avoids deadlock
    setTimeout(() => this.interactive.destroy(), 0);
    return false;
  }
}
